### DISCUSSION TITLE
Is there a way to add a data visualization dashboard into Next.JS?

### DISCUSSION THREAD

**Discussion by NextFoundArtistDuc** [vercel/next.js#71388] [Help]
### Summary

Do we have a package that could add a data visualization dashboard? 

I am thinking of connecting it with a MySQL database. I am building a movie streaming project that should be able to collect the number of views on a certain video, clicks on a video, and how long a user watches a video. I was wondering if we could create a data visualization dashboard on a webpage in Next.js that shows these metrics and kpis.

**Comment by madfcat** [comment]
Let me assist you with your problem.

There are several options to create a data visualization dashboard in a Next.js application. To connect with a MySQL database and display metrics such as the number of views, clicks, and watch time for your movie streaming project, you can use various libraries to create charts and dashboards.

Here’s a step-by-step breakdown of how you can achieve this:

## 1 . Connect to MySQL Database
To connect Next.js to your MySQL database, you can use an ORM (Object-Relational Mapper) like Prisma or a SQL query builder like Sequelize. Both provide easy ways to interact with your database.

- **Prisma**: A powerful ORM that works well with Next.js for handling data fetching on both server and client sides.
- **Sequelize**: A promise-based ORM for Node.js that can handle SQL databases.

**Example of using Prisma:**

1. Install **Prisma**:

```
npm install @prisma/client
```

1. Setup Prisma schema with your MySQL connection in `prisma/schema.prisma`

```
datasource db {
  provider = "mysql"
  url      = env("DATABASE_URL")
}

model Video {
  id      Int      @id @default(autoincrement())
  title   String
  views   Int
  clicks  Int
  watchTime Int
}
```

2. Fetch data in Next.js:
```
import { PrismaClient } from '@prisma/client';

const prisma = new PrismaClient();

export async function getServerSideProps() {
  const videos = await prisma.video.findMany();
  return {
    props: { videos },
  };
}
```

## 2. Data Visualization Libraries
Once the data is fetched, you can use various libraries to visualize it in your Next.js app. Here are some popular options:

**Recharts**: A composable charting library built on React.
**Chart.js**: A simple, flexible charting library that works well with React.**
**Victory**: A modular charting library that allows you to build interactive charts.

**Example with Recharts**:
1. Install Recharts:
```
npm install recharts
```

2. Create a component to display data:

```
import { LineChart, Line, XAxis, YAxis, CartesianGrid, Tooltip, Legend } from 'recharts';

const VideoMetricsChart = ({ data }) => {
  return (
    <LineChart
      width={500}
      height={300}
      data={data}
      margin={{
        top: 5, right: 30, left: 20, bottom: 5,
      }}
    >
      <CartesianGrid strokeDasharray="3 3" />
      <XAxis dataKey="title" />
      <YAxis />
      <Tooltip />
      <Legend />
      <Line type="monotone" dataKey="views" stroke="#8884d8" activeDot={{ r: 8 }} />
      <Line type="monotone" dataKey="clicks" stroke="#82ca9d" />
      <Line type="monotone" dataKey="watchTime" stroke="#ffc658" />
    </LineChart>
  );
};

export default VideoMetricsChart;
```

In this example, data would be an array of objects representing the metrics for each video, such as:
```
[
  { "title": "Movie 1", "views": 500, "clicks": 300, "watchTime": 1200 },
  { "title": "Movie 2", "views": 200, "clicks": 150, "watchTime": 800 }
]
```

## 3. Integrating a Dashboard UI Framework

To make the dashboard more user-friendly, you can use dashboard UI components:

- **Material UI**: A popular React UI framework that offers a wide range of components.
- **Ant Design**: Another great UI library that provides a lot of pre-built components for creating dashboards.

**Example with Material UI:**

1. Install Material UI:
```
npm install @mui/material @emotion/react @emotion/styled
```
2. Use Material UI components in your dashboard:
```
import { Container, Grid, Paper } from '@mui/material';
import VideoMetricsChart from './VideoMetricsChart';

const Dashboard = ({ data }) => {
  return (
    <Container maxWidth="lg">
      <Grid container spacing={3}>
        <Grid item xs={12}>
          <Paper>
            <h2>Video Metrics</h2>
            <VideoMetricsChart data={data} />
          </Paper>
        </Grid>
      </Grid>
    </Container>
  );
};

export default Dashboard;
```

## 4. Fetching Data from MySQL in Next.js API Routes

You can create API routes in Next.js to fetch data from your MySQL database and send it to your frontend components. This allows you to keep the data fetching logic separate from the rendering logic.

Example of an API route in `pages/api/videos.js`:
```
import { PrismaClient } from '@prisma/client';

const prisma = new PrismaClient();

export default async function handler(req, res) {
  const videos = await prisma.video.findMany();
  res.status(200).json(videos);
}
```

In your component, you can fetch this data using getServerSideProps or useEffect:
```
export async function getServerSideProps() {
  const res = await fetch('http://localhost:3000/api/videos');
  const videos = await res.json();

  return { props: { videos } };
}
```

## 5. Putting It All Together

Now, you can combine the data fetching, charting, and UI components into a single dashboard page that visualizes the video metrics.

Here’s a summary of steps:
- Use Prisma or Sequelize to connect to your MySQL database and fetch metrics.
- Use Recharts or Chart.js to display the data in charts.
- Use Material UI or Ant Design to create a polished dashboard layout.
- Fetch data using Next.js API routes or directly with getServerSideProps for server-side rendering.

This setup will allow you to build a comprehensive data visualization dashboard in Next.js to track and display KPIs for your movie streaming project.

**Reply by duchuynh** [reply]
Thanks so much! I was able to create a vibrant frontend of the dashboard so far using recharts.
